### PR TITLE
Torch pyg pl>=2.0

### LINKED ISSUE #477: Make GraphNeT compatible with pytorch-lightning >= 2.0
Currently there are some breaking changes in pytorch-lightning 2.0 which mean that GraphNeT doesn't work with it (see #470). It would be nice to be compatible with this newest version.

### LINKED ISSUE #540: Update installation to use torch, PyG, lightning >= 2 (should be much simpler for PyG)

### PR DESCRIPTION
**Pull request by AMHermansen**
This PR aims to make graphnet compatible with the newest versions of pytorch, pytorch lightning, and pytorch geometric. This comes at the cost of no longer being compatible with the older versions of those packages.

Currently the PR introduces a bug, in which the models are not saving due to a pickle error. But besides that, it is running on the latest versions of pytorch, pytorch lightning, and pytorch geometric.

Closes  #477
Closes #540

Update. The Pickle error has been fixed in a later commit.

### CODE REVIEW

File: src/graphnet/models/model.py [hunk 1]
```diff
@@ -39,6 +39,7 @@ def _construct_trainers(
         log_every_n_steps: int = 1,
         gradient_clip_val: Optional[float] = None,
         distribution_strategy: Optional[str] = "ddp",
+        inference_strategy: Optional[str] = "auto",
```
Comment (RasmusOrsoe) [inline src/graphnet/models/model.py:42]: I'm not a fan of introducing this argument.  Also, as it it's written now, the argument is not passed to the method - effectively hardcoding it to `"auto"`. 

I think a better solution would be a small refactor, where `_construct_trainers` is changed to

```
def _construct_trainer(
        self,
        max_epochs: int = 10,
        gpus: Optional[Union[List[int], int]] = None,
        callbacks: Optional[List[Callback]] = None,
        ckpt_path: Optional[str] = None,
        logger: Optional[LightningLogger] = None,
        log_every_n_steps: int = 1,
        gradient_clip_val: Optional[float] = None,
        distribution_strategy: Optional[str] = "ddp",
        **trainer_kwargs: Any,
    ) -> None:

        if gpus:
            accelerator = "gpu"
            devices = gpus
        else:
            accelerator = "cpu"
            devices = 1

        trainer = Trainer(
            accelerator=accelerator,
            devices=devices,
            max_epochs=max_epochs,
            callbacks=callbacks,
            log_every_n_steps=log_every_n_steps,
            logger=logger,
            gradient_clip_val=gradient_clip_val,
            strategy=distribution_strategy,
            ckpt_path = ckpt_path,
            **trainer_kwargs,
        )
   return trainer
```

This way, the function returns a single trainer based on a single set of `Trainer` arguments. We can then call this whenever the user interacts with this functionality through `Model`. So in `model.fit` we adjust:

```
def fit(
        self,
        train_dataloader: DataLoader,
        val_dataloader: Optional[DataLoader] = None,
        *,
        max_epochs: int = 10,
        gpus: Optional[Union[List[int], int]] = None,
        callbacks: Optional[List[Callback]] = None,
        ckpt_path: Optional[str] = None,
        logger: Optional[LightningLogger] = None,
        log_every_n_steps: int = 1,
        gradient_clip_val: Optional[float] = None,
        distribution_strategy: Optional[str] = "ddp",
        **trainer_kwargs: Any,
    ) -> None:
        """Fit `Model` using `pytorch_lightning.Trainer`."""
        # Checks
        if callbacks is None:
            callbacks = self._create_default_callbacks(
                val_dataloader=val_dataloader,
            )
        elif val_dataloader is not None:
            callbacks = self._add_early_stopping(
                val_dataloader=val_dataloader, callbacks=callbacks
            )

        self.train(mode=True)
        trainer = self._construct_trainer(
            max_epochs=max_epochs,
            gpus=gpus,
            callbacks=callbacks,
            ckpt_path=ckpt_path,
            logger=logger,
            log_every_n_steps=log_every_n_steps,
            gradient_clip_val=gradient_clip_val,
            distribution_strategy=distribution_strategy,
            **trainer_kwargs,
        )

        try:
            trainer.fit(
                self, train_dataloader, val_dataloader, ckpt_path=ckpt_path
            )
        except KeyboardInterrupt:
            self.warning("[ctrl+c] Exiting gracefully.")
            pass

```

and in `Model.predict`:

 ```
def predict(
        self,
        dataloader: DataLoader,
        gpus: Optional[Union[List[int], int]] = None,
        distribution_strategy: Optional[str] = "auto",
    ) -> List[Tensor]:
        """Return predictions for `dataloader`.

        Returns a list of Tensors, one for each model output.
        """
        self.train(mode=False)

        callbacks = self._create_default_callbacks(
                val_dataloader=None,
            )
        
        inference_trainer = self._construct_trainer(gpus=gpus, 
                                                    distribution_strategy=distribution_strategy, 
                                                    callbacks = callbacks)

        predictions_list = inference_trainer.predict(self, dataloader)
        assert len(predictions_list), "Got no predictions"

        nb_outputs = len(predictions_list[0])
        predictions: List[Tensor] = [
            torch.cat([preds[ix] for preds in predictions_list], dim=0)
            for ix in range(nb_outputs)
        ]

        return predictions
```

This _should_ also mean that there no references to `DataLoader` in `Model`. Hopefully this removes your pickle error.

**Review by RasmusOrsoe — CHANGES_REQUESTED** [review]
Hey @AMHermansen !

Thank you for giving this a shot. I have added comments.

I assume that the pickle error you mention is related to references to the Dataset/DataLoader in Model. One of my comments will likely remove this. Let me know what you think.

File: src/graphnet/data/dataset.py [hunk 2]
```diff
@@ -591,6 +591,12 @@ def _create_graph(
             add_these_to_graph.append(node_truth_dict)
         for write_dict in add_these_to_graph:
             for key, value in write_dict.items():
+                if value is None:
```
[the remark this answers is not in the document]
Comment (AMHermansen) [reply src/graphnet/data/dataset.py:594]: I was having an issue, where a None variable was passed to the `_create_graph`  method, I'm not sure if this is related to the upgrade, or the general framework. I'll try to see if I can recreate the bug, and make a bug report regarding it.

File: src/graphnet/models/model.py (hunk 1 above)
Comment (AMHermansen) [reply src/graphnet/models/model.py:42]: Thank you for your feedback. I was in doubt as to how we would want to implement a workaround for the lack of an inference strategy. I like the refactoring, and I will get back and see if this resolves the Pickle bug.

**Comment by RasmusOrsoe** [comment]
@AMHermansen 

Could you try to update the requirement specifications in `graphnet/requirements` such that they reflect the changes you've made here? I think that will resolve the issues with the automated checks failing for this PR.

**Comment by AMHermansen** [comment]
@RasmusOrsoe  It seems that there are no versions of torch>=2.0 which are compatible with python 3.7. Since this PR is already introducing backwards compatible breaking changes should we stop support for python 3.7?

File: src/graphnet/data/dataset.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (RasmusOrsoe) [reply src/graphnet/data/dataset.py]: Thanks. Issue is closed with #550

**Comment by RasmusOrsoe** [comment]
@AMHermansen Sorry for the wait. I spoke with IceCube IT and they do indeed have CVMFS with python 3.8 so I think we can go ahead and drop 3.7 without loss of deployment functionality.

**Comment by RasmusOrsoe** [comment]
Hey @AMHermansen!

I _think_ we can remove the 3.7 build by simply removing 3.7 from the list [here](https://github.com/graphnet-team/graphnet/blob/62db8c9126d30ebf3d181f819375950c021f5f8b/.github/workflows/build.yml#L76). Could you try that?

**Comment by AMHermansen** [comment]
Hello @RasmusOrsoe ,
I've removed python 3.7 from the build.yml, and the only thing that is failing is the installation in IceTray. I don't have enough experience to figure out what is causing the installation to fail for IceTray, while the regular installations are working as expected.
The new `_construct_trainer` also violates the check from codeclimate, but it is effectively the same as the previous method.
Do you have any insight into these issues?

**Comment by RasmusOrsoe** [comment]
@AMHermansen  Thank you! Now we're close to getting this through. Steps are below:


1. If you go to [this line](https://github.com/graphnet-team/graphnet/blob/62db8c9126d30ebf3d181f819375950c021f5f8b/.github/workflows/build.yml#L50) and add `pip install --upgrade psutil # Original version from IceTray Environment incompatible` below, the IceTray test should work.
2. Let's ignore codeclimate this time.
3. Please try to run the example scripts and let me know if there are any issues; if the pickle error persists then please let me know also.

**Comment by AMHermansen** [comment]
@RasmusOrsoe Thank you for the quick reply.
I've manged to successfully run all the training scripts locally. The refactoring of `_construct_trainer` did indeed solve the Pickle error that I was having earlier.

**Review by RasmusOrsoe — APPROVED** [review]
@AMHermansen Well done! You can now merge!
